libs/@blockprotocol/graph/src/shared/stdlib/subgraph/builder.ts

```diff
@@ -151,3 +154,84 @@ export const buildSubgraph = <Temporal extends boolean>(
 
   return subgraph;
 };
+
+/**
+ * Looking to build a subgraph? You probably want {@link buildSubgraph} from `@blockprotocol/graph/stdlib`
+ *
+ * This function will infer/add edges to a subgraph based on the vertices that are already present.
+ * The {@link Subgraph} must not have any edges already present, as this would result in invalid state.
+ * It will add edges for:
+ *  - data type vertices
+ *  - property type vertices
+ *  - entity type vertices
+ *  - entity vertices
+ *
+ * This operation MUTATES the given {@link Subgraph} - you should know why you need to do it.
+ *
+ * @param {Subgraph} subgraph – the subgraph to mutate by adding edges
+ */
+export const inferSubgraphEdges = <Temporal extends boolean>(
+  subgraph: Subgraph<Temporal, EntityRootType<Temporal>>,
+): void => {
+  // Construct object with vertex ids for each vertex kind
+  const vertexIds = typedEntries(subgraph.vertices).reduce(
+    (
+      acc: {
+        dataTypeVertexIds: OntologyTypeVertexId[];
+        propertyTypeVertexIds: OntologyTypeVertexId[];
+        entityTypeVertexIds: OntologyTypeVertexId[];
+        entityVertexIds: EntityVertexId[];
+      },
+      [baseId, entry],
+    ) => {
+      const revisionIds = typedKeys(entry);
+      let firstEntry = null;
+      if (revisionIds[0] && revisionIds[0] in entry) {
+        firstEntry = entry[revisionIds[0]];
+      }
```

Comment:
I genuinely found this quite hard to read and it took me a few goes to understand it :(

Maybe something like this?
```suggestion
      const sampleVertex = typedValues(entry).pop();

      const revisionIds = typedKeys(entry);
```

or if you're concerned about `Object.values` allocating a new array
```suggestion
      const revisionIds = typedKeys(entry);
      
      const sampleVertexRevisionId = revisionIds[0];
      const sampleVertex = sampleVertexRevisionId ? entry[sampleVertexRevisionId]! : null;
      }
```

Unless that second suggestion doesn't work? Bit confused about why you're having to do the `in` check, you should be able to assert it's non-null (considering you just checked it) but I didn't think you'd have to assert it's actually a valid index argument

Reply:
Alternatively we just rewrite this to be over `typedEntries` in the first place, and map the `newVertexIds` from that

Reply:
I commited this because it was so succinct 72616b0d6735ff341c47a4a14c4c5f8b3c19ff42
If you dislike it too much we can change it, but the template string actually fully type checks on the `kinds` we define and errors if anything is wrong for the types we've set. What do you think? :laughing:

Reply:
I thought I'd dislike it a lot more than I did, but upon looking at the plain file without the diff, I think it's fairly reasonable.

If it's properly type checked, I'm good with it :)